### PR TITLE
Modification for Arduino Core STM32

### PR DESCRIPTION
**Pull request by technik-gegg**
On the Arduino Core STM32 framework pin definitions have moved from uint8_t to uint32_t in order to reflect higher pin numbers on some of the newer MCUs available. 
Using uint8_t for storing pin numbers internally may cause cut offs and thus, such pins will not be adressed and function correctly.

Hence I've introduced a new typedef "pin_t" which will be set to uint32_t on the Adruino Core STM32 but stays as it is (uint8_t) on other frameworks:

```cpp
#if !defined(STM32_CORE_VERSION)
  typedef uint8_t     pin_t;
#else
  typedef uint32_t    pin_t;
#endif
```

I've also added a workaround for the missing ATOMIC_BLOCK in non AVR architectures, I've found on Stackoverflow.

### CODE REVIEW

File: src/SoftWire.h [hunk 1]
```diff
@@ -7,6 +7,32 @@
...
+//    If your project or library bases on/supports the Arduino Core STM32 framework,
+//    I'd highly recommend using the same construct as shown below in every location
+//    of your source code where you either store or return pin numbers, just to be future proof.
+//
+//    Using 32 bit for storing pin numbers might seem a bit overkill but one
+//    never knows what the future brings and eventually, using a specific type
+//    allows to adopt your source code quickly for a different framework without the 
+//    hassle of going through the code and changing/casting those pin definitions.
+//
+#if !defined(STM32_CORE_VERSION)
```
Comment (stevemarple) [inline src/SoftWire.h:30]: This can be simplified to remove the logical NOT and swap the if/else blocks.

File: src/SoftWire.h [hunk 2]
```diff
@@ -7,6 +7,32 @@
 #include <stdint.h>
 #include <AsyncDelay.h>
 
+//
+// 2022-04-13 by Technik Gegg
```
Comment (stevemarple) [inline src/SoftWire.h:11]: This is not necessary - it's just adds noise. git handles contributor and date information, and ensures it is correct.

File: src/SoftWire.h (hunk 1 above)
Comment (technik-gegg) [reply src/SoftWire.h]: So much simpler now ;o)

File: src/SoftWire.h (hunk 2 above)
Comment (technik-gegg) [reply src/SoftWire.h]: Removed

File: src/SoftWire.cpp [hunk 3]
```diff
@@ -4,8 +4,58 @@
...
+  }
+
+  static inline void __int_restore_irq(int *primask) {
+    if (!(*primask)) {
+      __enable_irq();
+    }
+  }
+  #define CRITICAL_SECTION for (int primask_save __attribute__((__cleanup__(__int_restore_irq))) = __int_disable_irq(), __ToDo = 1; __ToDo; __ToDo = 0)
+#else
+  #error Unsupported controller architecture
```
Comment (stevemarple) [inline src/SoftWire.cpp:58]: Doesn't this completely break the library for non-AVR and non-STM32 users, like ESP32? Issuing a pragma message seems more in keeping with the current library behaviour.

Comment (technik-gegg) [reply src/SoftWire.cpp:58]: What would you suggest?
Assigning an empty #define CRITICAL_SECTION and adding a warning instead to let the user know there might be a problem?

**Comment by technik-gegg** [comment]
> This is a useful set of changes but as there are two different changesets here:
> 
> 1. Fixing the pin type
> 2. Fixing the missing ATOMIC_BLOCK operation
> 
> I'd like to see a separate pull request for each.
> 
Ok, removed the ATOMIC_BLOCK changes from this PR.

> For the pin type you mentioned that only the newer STM32 frameworks use `uint32_t`. Rather than assuming a specific type is it possible to automatically use the same type as the framework does, probably by using [`decltype`](https://en.cppreference.com/w/cpp/language/decltype) ? Maybe something like this would work?
> 
> ```
> typedef decltype(some_pin_variable) pin_t;
> ```
> 
> Ideally this would be used for all Arduino cores, not just STM32, but I don't think there is a consistent way this can be applied. AVR cores (older ones anyway) use `#define` for some pins so restricting this change to within a `#ifdef` block is safest.

I haven't found anything in Arduino.h that would qualify for this notation.

File: src/SoftWire.cpp (hunk 3 above)
Comment (stevemarple) [reply src/SoftWire.cpp]: How about the conditional logic part starts with 

~~~
#ifdef CRITICAL_SECTION
// User has created their own workaround. Do nothing more here.

#elif 

// Your CRITICAL_SECTION stuff goes here

#else
  #pragma message "Your controller architecture is missing interrupt protection"
#endif
~~~

The first part means users can somehow add their own protection method externally to SoftWire. The last part warns users there is a problem. In production code I often have the compiler treat warnings as errors. However a library user cannot change their code to fix the library so I'd prefer a pragma message here. I realise that a `#define #CRITICAL_SECTION` could be used to silence the program/warning but the Arduino IDE doesn't provide a method for users to define macros for all compilation units, and I gave up fighting that battle long ago.

Comment (technik-gegg) [reply src/SoftWire.cpp]: Sounds like a feasible solution to me.

**Comment by stevemarple** [comment]
I've locally taken all of these commits, squashed them into one and compared to `master`. There are still changes to ATOMIC_BLOCK/CRITICAL section. I want a single changeset (and ideally just one commit, so I can see exactly what the change is) that adds the `pin_t` typedef.

I want to see the ATOMIC_BLOCK in a separate pull-request as this fixes a separate issue.

**Pull request closed without merging**